rust/operator-binary/src/discovery.rs

```diff
@@ -31,180 +26,126 @@ pub enum Error {
-        ..
-    }) = svc.spec.as_ref()
-    {
-        if service_type == &ServiceType::NodePort.to_string() {
-            discovery_configmaps.push(build_discovery_configmap(
-                &format!("{}-nodeport", name),
-                owner,
-                hive,
-                resolved_product_image,
-                chroot,
-                nodeport_hosts(client, svc, HIVE_PORT_NAME).await?,
-            )?);
-        }
-    }
-
     Ok(discovery_configmaps)
 }
 
 /// Build a discovery [`ConfigMap`] containing information about how to connect to a certain
 /// [`v1alpha1::HiveCluster`].
 ///
-/// `hosts` will usually come from the cluster role service or [`nodeport_hosts`].
+/// Data is coming from the [`Listener`] objects. Connection string is only build by [`build_listener_connection_string`].
 fn build_discovery_configmap(
     name: &str,
     owner: &impl Resource<DynamicType = ()>,
     hive: &v1alpha1::HiveCluster,
+    hive_role: HiveRole,
     resolved_product_image: &ResolvedProductImage,
     chroot: Option<&str>,
-    hosts: impl IntoIterator<Item = (impl Into<String>, u16)>,
+    listener: Listener,
 ) -> Result<ConfigMap, Error> {
-    let mut conn_str = hosts
-        .into_iter()
-        .map(|(host, port)| format!("thrift://{}:{}", host.into(), port))
-        .collect::<Vec<_>>()
-        .join("\n");
+    let mut discovery_configmap = ConfigMapBuilder::new();
+
+    discovery_configmap.metadata(
+        ObjectMetaBuilder::new()
+            .name_and_namespace(hive)
+            .name(name)
+            .ownerreference_from_resource(owner, None, Some(true))
+            .with_context(|_| ObjectMissingMetadataForOwnerRefSnafu {
+                hive: ObjectRef::from_obj(hive),
+            })?
+            .with_recommended_labels(build_recommended_labels(
+                hive,
+                &resolved_product_image.app_version_label,
+                &hive_role.to_string(),
+                "discovery",
+            ))
+            .context(MetadataBuildSnafu)?
+            .build(),
+    );
+
+    discovery_configmap.add_data(
+        "HIVE".to_string(),
+        build_listener_connection_string(listener, &hive_role.to_string(), chroot)?,
+    );
+
+    discovery_configmap
+        .build()
+        .with_context(|_| DiscoveryConfigMapSnafu {
+            obj_ref: ObjectRef::from_obj(hive),
+        })
+}
+
+// Builds the connection string with respect to the listener provided objects
+fn build_listener_connection_string(
+    listener_ref: Listener,
+    role: &String,
+    chroot: Option<&str>,
+) -> Result<String, Error> {
+    // We only need the first address corresponding to the role
+    let listener_address = listener_ref
+        .status
+        .and_then(|s| s.ingress_addresses?.into_iter().next())
+        .context(RoleGroupListenerHasNoAddressSnafu { role })?;
+    let mut conn_str = format!(
+        "thrift://{}:{}",
+        listener_address.address,
+        listener_address
```

Comment:
We should avoid using unnamed placeholders.

```suggestion
        "thrift://{address}:{port}",
        address = listener_address.address,
        port = listener_address
```